plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c

```diff
@@ -0,0 +1,261 @@
+}
+
+static void
+fu_uefi_dbx_snapd_notifier_finalize(GObject *object)
+{
+	FuUefiDbxSnapdNotifier *self = FU_UEFI_DBX_SNAPD_NOTIFIER(object);
+	curl_slist_free_all(self->req_hdrs);
+	curl_easy_cleanup(self->curl_template);
+	G_OBJECT_CLASS(fu_uefi_dbx_snapd_notifier_parent_class)->finalize(object);
+}
+
+static void
+fu_uefi_dbx_snapd_notifier_class_init(FuUefiDbxSnapdNotifierClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS(klass);
+	object_class->finalize = fu_uefi_dbx_snapd_notifier_finalize;
+}
+
+/* see CURLOPT_WRITEFUNCTION(3) */
+static size_t
+fu_uefi_dbx_snapd_notifier_rsp_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
+{
+	GByteArray *bufarr = (GByteArray *)userdata;
+	gsize sz = size * nmemb;
+	g_byte_array_append(bufarr, (const guint8 *)ptr, sz);
+	return sz;
+}
+
+static gboolean
+fu_uefi_dbx_snapd_notifier_simple_req(FuUefiDbxSnapdNotifier *self,
+				      const char *endpoint,
+				      const char *data,
+				      gsize len,
+				      GError **error)
+{
+	CURL *curl = NULL;
+	CURLcode res = -1;
+	glong status_code = 0;
+	g_autofree gchar *endpoint_str = NULL;
+	g_autoptr(GByteArray) rsp_buf = g_byte_array_new();
+
+	/* duplicate a preconfigured curl handle */
+	curl = curl_easy_duphandle(self->curl_template);
+
+	endpoint_str = g_strdup_printf("http://localhost%s", endpoint);
+
+	curl_easy_setopt(curl, CURLOPT_URL, endpoint_str);
+
+	g_debug("snapd simple request to %s with %" G_GSIZE_FORMAT " bytes of data",
+		endpoint_str,
+		len);
+	g_debug("request data: '%s'", data);
+	curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, len);
+	curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+
+	/* collect response for debugging */
+	curl_easy_setopt(curl, CURLOPT_WRITEDATA, rsp_buf);
+	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, fu_uefi_dbx_snapd_notifier_rsp_cb);
+
+	/* TODO: disable verbose or provide a handler */
+	curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+
+	res = curl_easy_perform(curl);
+	if (res != CURLE_OK) {
+		/* TODO inspect the error */
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "failed to communicate with snapd: %s",
+			    curl_easy_strerror(res));
+		g_message("failed to notify snapd: %s", curl_easy_strerror(res));
+		return FALSE;
+	}
+
+	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status_code);
+	curl_easy_cleanup(curl);
+
+	if (status_code == 404) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "snapd notification endpoint not supported by snapd API");
+		return FALSE;
+	}
+
+	if (status_code != 200) {
+		g_autofree gchar *rsp = NULL;
+		if (rsp_buf->len > 0) {
+			/* make sure the response is printable */
+			rsp = fu_strsafe((const char *)rsp_buf->data, rsp_buf->len + 1);
+		}
+
+		/* TODO check whether the response is even printable? */
+		g_message("snapd request failed with status %ld, response: %s",
+			  (glong)status_code,
+			  rsp != NULL ? rsp : "<none>");
```

Comment:
Actually it may make sense to keep this one as it shows the exact error response and it only appears if the status is really something else than 200 or 404.

Reply:
changed to info